Fixes to work with current API (2.1.9) and OS

sirkus7 opened the pull request:
Open Sprinkler API has changed, causing errors in opsiLCD.py. Updated
to work with v.2.1.9.

Also, fixed the network IP detection. Previous method of "hostname -I"
returned all addresses, including IPv6, with no guaranteed order. Now
uses the netifaces library and a defined variable for specifying which
interface will be used (wlan, eth0, etc.) The downside is that it
requires installation of netifaces. The upside is that this approach 
should be relatively stable and future proofed from OS changes. 
README updated to reflect library install requirement.

sirkus7 committed:
Fixes to work with current API and OS

Open Sprinkler API has changed, causing errors in opsiLCD.py. Updated
to work with v.2.1.9.

Also, fixed the network IP detection. Previous method of "hostname -I"
returned all addresses, including IPv6, with no guaranteed order. Now
uses the netifaces library and a defined variable for specifying which
interface will be used (wlan, eth0, etc.) The downside is that it
requires installation of netifaces. The upside is that it should be
a bit more future-proof. Here's hoping.

sirkus7 committed:
Updated link to API documentation, direct to docs

sirkus7 committed:
Fix README formatting for install instructions

wifi75 commented:
Sorry @sirkus7 tell me how to update the Ospilcd.py file?

sirkus7 commented:
@wifi75, for now it may be easiest for you to just clone my patched branch referenced above. Or you can visit my repo branch and just grab that file.

stanoba commented:
Hi @sirkus7. Thanks for fixing ospiLCD script. 👍
I'll have a look at it once I build my testing environment.

wifi75 commented:
Hello how i can update?

stanoba commented:
@wifi75 Go to @sirkus7 fork https://github.com/sirkus7/ospiLCD/tree/api-update-2.1.9 and follow his updated instructions.
Download his updated script https://github.com/sirkus7/ospiLCD/blob/api-update-2.1.9/ospiLCD.py instead of mine.

wifi75 commented:
@stanoba  @sirkus7 hello 
I follow this instruction: https://github.com/sirkus7/ospiLCD/blob/api-update-2.1.9/README.md
after set new script now this is my new error:
my network interface is wlan0

```
root@Opensprinkler:/home/pi# /home/pi/ospiLCD.py
Traceback (most recent call last):
  File "/home/pi/ospiLCD.py", line 52, in <module>
    ja = get_data(api_url)
  File "/home/pi/ospiLCD.py", line 48, in get_data
    variables = json.loads(data, object_hook=lambda d: namedtuple('X', d.keys())(*d.values()))
  File "/usr/lib/python2.7/json/__init__.py", line 352, in loads
    return cls(encoding=encoding, **kw).decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 380, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/home/pi/ospiLCD.py", line 48, in <lambda>
    variables = json.loads(data, object_hook=lambda d: namedtuple('X', d.keys())(*d.values()))
  File "/usr/lib/python2.7/collections.py", line 354, in namedtuple
    'keyword: %r' % name)
ValueError: Type names and field names cannot be a keyword: 'pass'
root@Opensprinkler:/home/pi#


```

sirkus7 committed:
Update README apt install command.

Python pip is not included in all Raspian images and must first be
installed. Added pip to the existing apt isntall command, and moved it
to the beginning to make it work smoothly.

sirkus7 committed:
Fix keyword conflict in JSON response.

The new OpenSprinkler API responds with a parameter called "pass" which
is a keyword and therefore cannot be used to create an object.
Fixed by replacing "pass" term with "password" prior to deserializing
the JSON response.

sirkus7 commented:
@wifi75 I tested on a new install and hit the same issue you did, so clearly I had some other differences on my system that allowed things to work. So thanks for pointing out that problem.
I found the problem and committed the fix.  Grab the new one and try it again.

sirkus7 committed:
Delete test file.

wifi75 commented:
@sirkus7 
do I buy the new file from here?
wget https://raw.githubusercontent.com/stanoba/ospiLCD/master/ospiLCD.py


after that operation same problem:

```
root@Opensprinkler:/home/pi# ./ospiLCD.py
Traceback (most recent call last):
  File "./ospiLCD.py", line 52, in <module>
    ja = get_data(api_url)
  File "./ospiLCD.py", line 48, in get_data
    variables = json.loads(data, object_hook=lambda d: namedtuple('X', d.keys())(*d.values()))
  File "/usr/lib/python2.7/json/__init__.py", line 352, in loads
    return cls(encoding=encoding, **kw).decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python2.7/json/decoder.py", line 380, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "./ospiLCD.py", line 48, in <lambda>
    variables = json.loads(data, object_hook=lambda d: namedtuple('X', d.keys())(*d.values()))
  File "/usr/lib/python2.7/collections.py", line 354, in namedtuple
    'keyword: %r' % name)
ValueError: Type names and field names cannot be a keyword: 'pass'
```

sirkus7 commented:
@wifi75 That's the old version.  You can get the one I updated here: 
https://github.com/sirkus7/ospiLCD/blob/api-update-2.1.9/ospiLCD.py

wifi75 commented:
@sirkus7  ok yes now the script work but display it is black, where is on opensprinkler setting the light level? 
once there was

sirkus7 commented:
@wifi75 Haha, I honestly don't know where that setting can be found. I've only recently installed opensprinkler, and I haven't found the light level setting in the UI. It's possible it has been removed for the RPi version.  

For my purposes, I turned the backlight on permanently because I couldn't find that feature.  If that would be okay for you as well, you can add the following line of code at line 164 (in my version of the script) to turn the backlight on:
`backlight = True`

*[Update: I corrected the line above]*

wifi75 commented:
that 
lcd.backlight_enabled = True
in your ospi file ?

sirkus7 commented:
Yes. Though, let me correct something.  The line should be
``` python
backlight = True
```
So, If you've used my updated version of the ospiLCD.py file, and if you'd like the LCD backlight to be on all the time, then you can simply insert that one line of code at line into the ospiLCD.py file at line 164.  Just to show that line in context, the new line of code is placed right before the lcd device is created to display the information, so it would look like this in the script. 

```python 
backlight = True
lcd = i2c.CharLCD(i2c_expander=LCD_i2c_expander, address=LCD_i2c_address, port=1, cols=LCD_cols, rows=LCD_rows,
					dotsize=8, charmap='A02', auto_linebreaks=True, backlight_enabled=backlight)
```

The result will be that the backlight is always enabled.  Hope that helps.

wifi75 commented:
nice now work!
thank you
